# Notebook 13 – Missing Values



# 1.isnull()

## Step 1 – Understand the Concept

`isnull()` is used to identify missing values in a Pandas DataFrame or Series.

A missing value means that some information is not available.

For example:

| Name | Age |
|---|---:|
| Arun | 22 |
| Priya | NaN |
| Kumar | 24 |

Here, Priya's Age is missing.

`isnull()` returns:
- `True` → value is missing
- `False` → value is not missing

It is one of the first methods used when checking the quality of a dataset.

## Step 2 – Demonstrate the Concept

### Simple Example

If a student's marks are missing, `isnull()` can identify that missing value.

### Real-World Example

A customer dataset may have missing phone numbers or email addresses.

### Business Example

A sales dataset may contain transactions where the sales amount is missing.

### AI/ML Use Case

Before training a machine learning model, we need to identify missing values in the features.

### Data Analytics Use Case

Analysts can use `isnull()` to understand which columns contain incomplete data.


## Step 3: Implement the Concept

In [1]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Arun", "Priya", "Kumar"],
    "Age": [22, None, 24],
    "Marks": [85, 90, None]
})

print(df.isnull())
print("\nMissing values in each column:")
print(df.isnull().sum())

    Name    Age  Marks
0  False  False  False
1  False   True  False
2  False  False   True

Missing values in each column:
Name     0
Age      1
Marks    1
dtype: int64


## Code Explanation

- `isnull()` checks every value in the DataFrame.
- It returns `True` when a value is missing.
- `isnull().sum()` counts the missing values in each column.
- This helps us understand how much missing data exists.
- In AI/ML, this step is useful before model training.
- In Data Analytics, it helps identify incomplete columns and records.



# 2. notnull()


## Step 1 – Understand the Concept

`notnull()` is used to identify values that are available and not missing.

It is the opposite of `isnull()`.

It returns:

- `True` → value is available
- `False` → value is missing

This method is useful when we want to select only records that contain valid values.

## Step 2 – Demonstrate the Concept

### Simple Example

If a student's marks are available, `notnull()` returns `True`.

### Real-World Example

A customer database can be filtered to find customers whose email addresses are available.

### Business Example

A company can select sales records where the sales amount is available.

### AI/ML Use Case

`notnull()` can be used to select records with available feature values for analysis.

### Data Analytics Use Case

Analysts can filter datasets to work only with records containing required information.


## Step 3: Implement the Concept

In [2]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Arun", "Priya", "Kumar"],
    "Age": [22, None, 24],
    "Marks": [85, 90, None]
})

print(df.notnull())

print("\nRows where Age is available:")
print(df[df["Age"].notnull()])

   Name    Age  Marks
0  True   True   True
1  True  False   True
2  True   True  False

Rows where Age is available:
    Name   Age  Marks
0   Arun  22.0   85.0
2  Kumar  24.0    NaN


## Code Explanation

- `notnull()` checks whether values are available.
- It returns `True` for non-missing values.
- `df["Age"].notnull()` creates a Boolean condition for the Age column.
- The DataFrame is filtered using this condition.
- In AI/ML, it can help select records with required data.
- In Data Analytics, it can be used to filter complete values from a column.



# 3.dropna()

## Step 1 – Understand the Concept

`dropna()` is used to remove rows or columns containing missing values.

For example, if a row contains a missing value, `dropna()` can remove that row.

By default, `dropna()` removes rows containing at least one missing value.

It is useful when missing records are not required or when the number of missing values is small.

## Step 2 – Demonstrate the Concept

### Simple Example

If one student's marks are missing, we can remove that student's row using `dropna()`.

### Real-World Example

A survey dataset may contain a few incomplete responses that are not useful for analysis.

### Business Example

A company may remove transaction records where important required information is missing.

### AI/ML Use Case

Rows with missing values may sometimes be removed before model training when the number of missing records is very small.

### Data Analytics Use Case

Analysts can remove incomplete records when missing information is not important for the analysis.


## Step 3: Implement the Concept

In [3]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Arun", "Priya", "Kumar", "Anitha"],
    "Age": [22, None, 24, 25],
    "Marks": [85, 90, None, 88]
})

clean_df = df.dropna()

print(clean_df)

     Name   Age  Marks
0    Arun  22.0   85.0
3  Anitha  25.0   88.0


## Code Explanation

- `dropna()` removes rows containing missing values.
- In this example, rows with missing Age or Marks are removed.
- `clean_df` contains only rows without missing values.
- `dropna()` should be used carefully because removing rows can reduce the amount of available data.
- In AI/ML, it may be suitable when only a small number of rows have missing values.
- In Data Analytics, it can help create a complete dataset for certain analyses.



# 4.fillna()

## Step 1 – Understand the Concept

`fillna()` is used to replace missing values with another value.

Instead of removing a row containing missing data, we can fill the missing value.

For example:

Missing Age → 25

We can also use statistical values such as:
- Mean
- Median
- Mode

The choice depends on the type of data and the reason for the missing value.

## Step 2 – Demonstrate the Concept

### Simple Example

If one student's age is missing, we can fill it with a suitable value.

### Real-World Example

A customer dataset may have missing ages. The missing values can be filled using an appropriate statistical value.

### Business Example

A company may fill missing sales values using a carefully selected business rule or statistical method.

### AI/ML Use Case

Missing numerical features can sometimes be replaced using the mean or median before model training.

### Data Analytics Use Case

Analysts can fill missing values to avoid losing useful records during analysis.


## Step 3: Implement the Concept

In [4]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Arun", "Priya", "Kumar", "Anitha"],
    "Age": [22, None, 24, 25],
    "Marks": [85, 90, None, 88]
})

df["Age"] = df["Age"].fillna(df["Age"].mean())
df["Marks"] = df["Marks"].fillna(df["Marks"].mean())

print(df)


     Name        Age      Marks
0    Arun  22.000000  85.000000
1   Priya  23.666667  90.000000
2   Kumar  24.000000  87.666667
3  Anitha  25.000000  88.000000


## Code Explanation

- `fillna()` replaces missing values with the specified value.
- `df["Age"].mean()` calculates the average Age.
- The missing Age is replaced with the mean Age.
- The same approach is used for Marks.
- This keeps the rows instead of removing them.
- In AI/ML, imputation helps prepare data for models that cannot handle missing values.
- In Data Analytics, it allows analysts to retain more records for analysis.



# 5. interpolate()

## Step 1 – Understand the Concept

`interpolate()` is used to estimate missing numerical values based on surrounding values.

Instead of simply filling a missing value with one fixed value, interpolation calculates a value based on the existing values around it.

For example:

10, 20, NaN, 40

The missing value can be estimated as:

30

Interpolation is especially useful when data follows an ordered sequence, such as time-series data.

## Step 2 – Demonstrate the Concept

### Simple Example

For the values:

10, 20, NaN, 40

interpolation can estimate the missing value as 30.

### Real-World Example

A temperature sensor may fail to record one reading. The missing temperature can be estimated from nearby readings.

### Business Example

A company may have a missing daily sales value in a time-series dataset and estimate it using nearby values.

### AI/ML Use Case

Interpolation can be used to fill missing values in time-dependent features before training a model.

### Data Analytics Use Case

Analysts can use interpolation to maintain continuous time-series data for trend analysis.


## Step 3: Implement the Concept

In [5]:
import pandas as pd

df = pd.DataFrame({
    "Day": [1, 2, 3, 4, 5],
    "Sales": [100, 120, None, 160, 180]
})

df["Sales"] = df["Sales"].interpolate()

print(df)

   Day  Sales
0    1  100.0
1    2  120.0
2    3  140.0
3    4  160.0
4    5  180.0


## Code Explanation

- `interpolate()` estimates missing numerical values using surrounding values.
- The missing Sales value between 120 and 160 is estimated as 140.
- This method is useful when the data has an order or sequence.
- In AI/ML, interpolation can help prepare continuous time-series features.
- In Data Analytics, it is useful for filling gaps in time-series data.



# 6. Missing Value Handling Strategies

## Step 1 – Understand the Concept

Missing value handling means deciding what to do when a dataset contains unavailable values.

There is no single method that is correct for every dataset.

Common strategies include:

1. Remove rows using `dropna()`
2. Remove columns when too many values are missing
3. Fill values using `fillna()`
4. Use mean, median, or mode
5. Use `interpolate()` for suitable numerical or time-series data
6. Keep missing values when the missingness itself has meaning

The correct strategy depends on:
- Amount of missing data
- Type of column
- Business meaning
- Data distribution
- Machine learning requirements

## Step 2 – Demonstrate the Concept

### Simple Example

If only one value is missing, we may fill it instead of removing the entire row.

### Real-World Example

If a sensor occasionally misses a reading, interpolation may be suitable.

### Business Example

If a customer phone number is missing, replacing it with the average value does not make sense. The missing value may need to remain missing or be handled using a business rule.

### AI/ML Use Case

Before training a model, missing values must be handled appropriately because many ML algorithms cannot work directly with missing values.

### Data Analytics Use Case

Analysts choose a missing-value strategy based on how much data is missing and how important the affected column is.

### Important Point

We should not automatically replace every missing value with zero.

The correct strategy depends on what the missing value represents.


## Step 3 – Implement the Concept

In [6]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Arun", "Priya", "Kumar", "Anitha", "Rahul"],
    "Age": [22, None, 24, 25, None],
    "Marks": [85, 90, None, 88, 76]
})

print("Original Data:")
print(df)

# Check missing values
print("\nMissing Values:")
print(df.isnull().sum())

# Fill numerical missing values using median
df["Age"] = df["Age"].fillna(df["Age"].median())
df["Marks"] = df["Marks"].fillna(df["Marks"].median())

print("\nAfter Handling Missing Values:")
print(df)

Original Data:
     Name   Age  Marks
0    Arun  22.0   85.0
1   Priya   NaN   90.0
2   Kumar  24.0    NaN
3  Anitha  25.0   88.0
4   Rahul   NaN   76.0

Missing Values:
Name     0
Age      2
Marks    1
dtype: int64

After Handling Missing Values:
     Name   Age  Marks
0    Arun  22.0   85.0
1   Priya  24.0   90.0
2   Kumar  24.0   86.5
3  Anitha  25.0   88.0
4   Rahul  24.0   76.0


## Code Explanation

- `isnull().sum()` identifies the number of missing values in each column.
- `median()` calculates the middle value of a numerical column.
- `fillna()` replaces the missing values with the calculated median.
- Median is often useful when numerical data contains extreme values.
- In AI/ML, appropriate missing-value handling is an important preprocessing step.
- In Data Analytics, it helps preserve useful records while preparing clean data.

### Choosing a Strategy

- Few missing rows → `dropna()` may be suitable.
- Numerical data → mean or median can be considered.
- Categorical data → mode or a meaningful category can be considered.
- Time-series data → `interpolate()` may be suitable.
- Too many missing values → consider removing the column after understanding its importance.